### PR TITLE
Remove uslep(100) from first call of cache->get in file handler

### PR DESCRIPTION
**Pull request by csthomas**
Pull Request for Issue #10706.
#### Summary of Changes

1) Eliminate usleep(100) when cache file does not exist.
2) Do not call `flock()` if the file does not exist, replace flag `r+b` to `c+b` in `fopen` function.
3) Do not allow to release lock, remember `$_fileopen` resource in $this->_locked_files variable.
4) At `unlock()` do not re-open cache file, but release lock only.

#### Testing Instructions

Code review.
Or standard test below.

**Before patch.**
1) Enable cache and set cache handler to File,
2) Replace line 309 in `libraries/joomla/cache/storage/file.php` ( https://github.com/joomla/joomla-cms/pull/10767/files#diff-056f1b3f5ee42d2b80e891d375789137L309 )
from:
``` php
usleep(100);
```

to 

``` php
usleep(100); JLog::add("Joomla call usleep!!!", JLog::WARNING, 'jerror');
```

3) Remove all files from `cache`. Clear cache.
4) Now go to front-end.
5) You should see "Joomla call usleep!!!"

**After patch** do the same but 
at point 5 you should not see "Joomla call usleep!!!"

### CODE REVIEW

**Comment by brianteeman** [comment]
Without any testing instructions this is unlikely to get tested very quickly<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10767">issues.joomla.org/joomla-cms/10767</a>.</sub>

**Comment by csthomas** [comment]
I have added a test instruction in first comment.

**Comment by brianteeman** [comment]
Thank you

File: libraries/joomla/cache/storage/file.php [hunk 1]
```diff
@@ -265,17 +265,15 @@ public function lock($id, $group, $locktime)
 
 		$looptime  = $locktime * 10;
 		$path      = $this->_getFilePath($id, $group);
-		$_fileopen = @fopen($path, "r+b");
+		$_fileopen = @fopen($path, "w+b");
```
Comment (csthomas) [inline libraries/joomla/cache/storage/file.php]: This line open or create a file (w+b), instead only open file (r+b) if exists.

File: libraries/joomla/cache/storage/file.php [hunk 2]
```diff
@@ -265,17 +265,15 @@ public function lock($id, $group, $locktime)
...
 
-		if ($_fileopen)
+		if (!$_fileopen)
 		{
-			$data_lock = @flock($_fileopen, LOCK_EX);
-		}
-		else
-		{
-			$data_lock = false;
+			return false;
```
Comment (csthomas) [inline libraries/joomla/cache/storage/file.php]: If file can not be open (disk full) then `return false;`

File: libraries/joomla/cache/storage/file.php [hunk 3]
```diff
@@ -265,17 +265,15 @@ public function lock($id, $group, $locktime)
...
 		}
 
+		$data_lock = @flock($_fileopen, LOCK_EX);
+
 		if ($data_lock === false)
```
Comment (csthomas) [inline libraries/joomla/cache/storage/file.php]: If the file has not existed then before this PR we go to the loop below.

**Comment by csthomas** [comment]
I updated instruction and code. Please test.

**Comment by csthomas** [comment]
@mbabker @zero-24 Can you take a look at the new test, a few build fails and what do you think about flock() function inconsistency?

**Comment by csthomas** [comment]
Feel free to test.

**Comment by franz-wohlkoenig** [comment]
I have tested this item :white_check_mark: successfully on c41eda06245e55e1655f89da6c18012593c96b8e<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10767">issues.joomla.org/tracker/joomla-cms/10767</a>.</sub>

**Comment by marrouchi** [comment]
I have tested this item :white_check_mark: successfully on b85ddcad5398141c326872b25ceab72af2585b3b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10767">issues.joomla.org/tracker/joomla-cms/10767</a>.</sub>

**Comment by csthomas** [comment]
Thanks for tests.

First test was done before 'resolve merge conflict' but 'conflicted' merged PR 13372 only change `"` to `'`.  
Can we add RTC as we have two success tests.

**Comment by jeckodevelopment** [comment]
@csthomas We should have another test, please, since we had new commits after the test of @franz-wohlkoenig.
If he can test it again, it would be great

**Comment by franz-wohlkoenig** [comment]
will do

**Comment by franz-wohlkoenig** [comment]
I have tested this item :white_check_mark: successfully on b85ddcad5398141c326872b25ceab72af2585b3b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10767">issues.joomla.org/tracker/joomla-cms/10767</a>.</sub>

**Comment by csthomas** [comment]
thanks

**Comment by dgt41** [comment]
I have tested this item :white_check_mark: successfully on b85ddcad5398141c326872b25ceab72af2585b3b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10767">issues.joomla.org/tracker/joomla-cms/10767</a>.</sub>

**Comment by dgt41** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10767">issues.joomla.org/tracker/joomla-cms/10767</a>.</sub>

**Comment by wilsonge** [comment]
Changing `r+b` to `c+b` appears to be responsible for https://github.com/joomla/joomla-cms/issues/15544

**Comment by csthomas** [comment]
I'm checking it

**Comment by wilsonge** [comment]
Thanks :)

**Comment by csthomas** [comment]
I see a few solution.
The reason is that the cache create a `lock` before call view method.
If view method generete 404 then process call error page and die.
Cache method does not get any response about failed view.
Cache file is empty and stay to the next requests.

1. First solution is dirty, because because use registered `shutdown` function to delete such cache file.
```diff
diff --git a/libraries/joomla/cache/storage/file.php b/libraries/joomla/cache/storage/file.php
index f2dba86e5f..c4d01e4ed7 100644
--- a/libraries/joomla/cache/storage/file.php
+++ b/libraries/joomla/cache/storage/file.php
@@ -361,6 +361,19 @@ class JCacheStorageFile extends JCacheStorage
                {
                        // Remember resource, flock release lock if you unset/close resource
                        $this->_locked_files[$path] = $_fileopen;
+
+                       $pointer =& $this->_locked_files[$path];
+
+                       // Delete file on unexpected die.
+                       $deleteMe = function () use ($pointer, $path)
+                       {
+                               if ($pointer)
+                               {
+                                       unlink($path);
+                               }
+                       };
+
+                       register_shutdown_function($deleteMe);
                }
```

2. Second solution is simple, does not delete such file but only check file size:
```diff
diff --git a/libraries/joomla/cache/storage/file.php b/libraries/joomla/cache/storage/file.php
index f2dba86e5f..ac0f54ccad 100644
--- a/libraries/joomla/cache/storage/file.php
+++ b/libraries/joomla/cache/storage/file.php
@@ -413,7 +413,7 @@ class JCacheStorageFile extends JCacheStorage
                {
                        $time = @filemtime($path);
 
-                       if (($time + $this->_lifetime) < $this->_now || empty($time))
+                       if (($time + $this->_lifetime) < $this->_now || empty($time) || filesize($path) === 0)
                        {
                                @unlink($path);

```
I will create a PR for above option 2.

Other solutions may require to catch each error as exception in the cache method and then delete the cache file or try to cache error response.

**Comment by demis-palma** [comment]
Hi @csthomas I have experienced another issue due to 'r+b' -> 'c+b' replacement.
Before this change, for a module included in an article with {loadmodule}, it was possible to call JCache::setCaching() to disable caching of the container article.
Now an empty file remains in the directory instead, which prevents subsequent view of the article until the cache expires.

Therefore I ask myself if 'c+b' is really necessary? What's the advantage of creating an empty file when you still don't need it?
Otherwise I'd suggest to restore the previous mode 'r+b'.

**Comment by csthomas** [comment]
About creating an empty file.
* The empty file is required to lock this file by `flock()`, to prevent other php processes to create the same cache file in the same time.

About your situation I understand that:
1. The cache is enabled.
2. There is a request for article view.
3. The cache library tries to find a cache file but there is not file.
4. The cache library creates an empty file and locks it.
5. The cache library calls `$view->display()`
6. The module loaded by {loadmodule} disables the cache by `JCache::setCaching(false)`
7. The cache library gets the data but does not store it.
8. The empty cache file left.
9. At the next request the empty cache file exists and there is the problem. 

Take a look at https://github.com/joomla/joomla-cms/pull/15592 it should resolve your problem.

**Comment by demis-palma** [comment]
@csthomas your description is so accurate that for a second, I was afraid that you control my PC with a malware. :stuck_out_tongue_winking_eye:

I understand the need for file locking, but this way we have changed so deeply the behaviour of the file cache, that the problem reported up to now are just the first in a long line that I'm sure will be reported.

I think that the current implementation has a major architectural problem that was already there before your PR, and that can be resumed in "do not lock for longer than is strictly necessary".
Since we currently:

1. lock the file 2 hours and a halt before we are ready to write data into it, **and**
2. give controls to third party components and modules in the meantime,

then it is reasonable to expect that something **will** go wrong.
Your attempt to fix it using the shutdown function is clever, but as you said, it's dirty.

To me it was badly designed before your PR, but the 'r+b' mode did prevent early lock. 
It's one of those cases where two errors compensates each other.
Having changed the mode to 'c+b' has highlighted the problem.

I just wanted to know if you are agree with that, and I'll start a full refactoring.

**Pull request merged**